### PR TITLE
GraphQL - reports

### PR DESCRIPTION
**Pull request by toyhammered**
# What

Add report functionality to GraphQL.

# Why

# Checklist

- [ ] All files pass Rubocop
- [ ] Any complex logic is commented
- [ ] Any new systems have thorough documentation
- [ ] Any user-facing changes are behind a feature flag (or: explain why they can't be)

- [ ] All the tests pass
- [ ] Tests have been added to cover the new code

### CODE REVIEW

File: app/graphql/types/report.rb [hunk 1]
```diff
@@ -0,0 +1,35 @@
...
+
+  field :status, Types::Enum::ReportStatus,
+    null: false,
+    description: 'The resolution status for this report'
+
+  field :explanation, String,
+    null: true,
+    description: <<~DESCRIPTION.squish
+      Additional information related to why the report was made.
+      The most relevant results will be at the top.
```
Comment (Reinachan) [inline app/graphql/types/report.rb]: I'm not sure what this line is for? How is relevance related to a report description? Is this line here by accident?

File: app/graphql/types/union/report_item.rb [hunk 2]
```diff
@@ -0,0 +1,5 @@
+class Types::Union::ReportItem < Types::Union::Base
+  description 'Objects which are Reportable'
+
+  possible_types Types::Post, Types::Comment, Types::MediaReaction, Types::Review
```
Comment (Reinachan) [inline app/graphql/types/union/report_item.rb:4]: Does this mean that you can fetch details about the post/comment/reaction/review through your query? Like doing something like this:

```gql
query reports {
  // stuff
  naughty {
    ... on Post {
      content
      user {
        username
      }
      // more stuff
    }
  }
}
```

**Review by Reinachan — APPROVED** [review]
Other than these two questions, LGTM. If the answer to the second question is no, then that would need to be addressed :)

File: app/graphql/types/report.rb (hunk 1 above)
Comment (toyhammered) [reply app/graphql/types/report.rb]: oops, will remove.

File: app/graphql/types/union/report_item.rb (hunk 2 above)
Comment (toyhammered) [reply app/graphql/types/union/report_item.rb:4]: yes

File: app/graphql/types/query_type.rb [hunk 3]
```diff
@@ -395,4 +395,36 @@ def franchises
   #   DESCRIPTION
   #   argument :title, String, required: true
   # end
+
+  field :reports, Types::Report.connection_type, null: true do
+    description 'All Reports in the Kitsu database'
+  end
+
+  def reports
+    ReportPolicy::Scope.new(context[:token], ::Report).resolve
```
Comment (NuckChorris) [inline app/graphql/types/query_type.rb]: You probably want to add a sort to this, even if it's only `.order(created_at: :desc)` 😅

File: app/graphql/types/review.rb [hunk 4]
```diff
@@ -0,0 +1,60 @@
...
+    description: 'Potentially migrated over from hummingbird.'
+
+  field :is_spoiler, Boolean,
+    null: false,
+    description: 'Does this review contain spoilers from the media',
+    method: :spoiler
+
+  field :deleted_at, GraphQL::Types::ISO8601DateTime,
+    null: true,
+    description: 'When this review was deleted'
```
Comment (NuckChorris) [inline app/graphql/types/review.rb]: This doesn't need to be exposed, deleted reviews aren't accessible via the API

File: app/graphql/types/review.rb [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+  field :likes, Types::Profile.connection_type,
+    null: false,
+    description: 'Users who liked this reaction'
```
Comment (NuckChorris) [inline app/graphql/types/review.rb]:
```suggestion
    description: 'Users who liked this review'
```

**Review by NuckChorris — APPROVED** [review]
LGTM!

**Pull request merged**